### ISSUE TITLE
Transaction not working in a Cluster Redis

### ISSUE THREAD

**Issue by ttecles** [alisaifee/coredis#118]
I am trying to run the transaction example you expose in the docs in a Redis Cluster but even if using a specific key for a specific node, I get an error cause the get returns the pipe instead of the value:

```python
async def client_side_incr(pipe):
    current_value = await pipe.get('OUR-SEQUENCE-KEY')
    next_value = int(current_value) + 1
    pipe.multi()
    await pipe.set('OUR-SEQUENCE-KEY', next_value)

await r.transaction(client_side_incr, 'OUR-SEQUENCE-KEY')
# [True]
```

### Error 
```
tests/integration/test_session.py:11 (test_session_can_save_a_key)
Traceback (most recent call last):
  File "/home/user/apps/product-images/tests/integration/test_session.py", line 23, in test_session_can_save_a_product_with_images
    await redis_client.transaction(client_side_incr, 'OUR-SEQUENCE-KEY')
  File "/home/joan/virtualenvs/product-images/lib/python3.10/site-packages/coredis/client/cluster.py", line 1058, in transaction
    func_value = await func(pipe)
  File "/home/user/apps/product-images/tests/integration/test_session.py", line 19, in client_side_incr
    next_value = int(current_value) + 1
TypeError: int() argument must be a string, a bytes-like object or a real number, not 'ClusterPipelineImpl'
```

**Comment by alisaifee** [comment]
You're right, that example is broken for cluster and in fact the implementation of `transaction` in cluster mode is pretty broken. I did some quick checks and it seems like it's been in this state even in the original [aredis](https://github.com/NoneGG/aredis) implementation from which coredis was forked.

This might require some effort to get working (if that's even viable since I assume the original author had difficulty completing the implementation and also, I see that transactions are deprecated in [redis-py](https://github.com/redis/redis-py/blob/master/redis/cluster.py#L704).

In the meanwhile:
1) I'll probably add some exceptions & update the docs to clearly point out that this doesn't work in cluster
2) what is the actual business logic you're trying to implement. Perhaps we can brainstorm on other ways to achieve it with coredis.

**Comment by ttecles** [comment]
> You're right, that example is broken for cluster and in fact the implementation of `transaction` in cluster mode is pretty broken. I did some quick checks and it seems like it's been in this state even in the original [aredis](https://github.com/NoneGG/aredis) implementation from which coredis was forked.
> 
> This might require some effort to get working (if that's even viable since I assume the original author had difficulty completing the implementation and also, I see that transactions are deprecated in [redis-py](https://github.com/redis/redis-py/blob/master/redis/cluster.py#L704).
> 
> In the meanwhile:
> 
> 1. I'll probably add some exceptions & update the docs to clearly point out that this doesn't work in cluster
> 2. what is the actual business logic you're trying to implement. Perhaps we can brainstorm on other ways to achieve it with coredis.

We have multiple clients doing CAS operations on the same key. First, we were using the WATCH but since it isn't possible on the client library (it is with a raw client, cause we are accessing just one key) we decide to go with a transaction. But again, the WATCH isn't working on a transaction with a cluster and `pipe.get` returns the value if it is a `Redis` client but a `ClusterPipelineImp` if it is a `RedisCluster` client. We tried the following:

### Client 1
```python
import asyncio

from coredis import RedisCluster


def get_redis_cluster_client():
    startup_nodes = [{'host': 'localhost', 'port': 6379}]
    return RedisCluster(startup_nodes=startup_nodes, decode_responses=True, protocol_version=2)


async def test_lock():
    client = get_redis_cluster_client()

    async def client_side_incr(pipe):
        pipe = await pipe.get('KEY')
        current_value = (await pipe.execute())[0]
        print(f'Initial: {current_value}')
        current_value = int(current_value or 0)
        await asyncio.sleep(10)
        await pipe.set('KEY', current_value + 1)

    await client.transaction(client_side_incr, 'KEY')

    final_value = await client.get('KEY')
    print(f'Final: {final_value}')


asyncio.run(test_lock())

```

### Client 2
```python
import asyncio
from coredis import RedisCluster


def get_redis_cluster_client():
    startup_nodes = [{'host': 'localhost', 'port': 6379}]
    return RedisCluster(startup_nodes=startup_nodes, decode_responses=True, protocol_version=2)


async def test_lock():
    client = get_redis_cluster_client()

    async def client_side_incr(pipe):
        pipe = await pipe.get('KEY')
        current_value = (await pipe.execute())[0]
        print(f'Initial: {current_value}')
        current_value = int(current_value or 0)
        await pipe.set('KEY', current_value + 1)

    await client.transaction(client_side_incr, 'KEY')

    final_value = await client.get('KEY')
    print(f'Final: {final_value}')


asyncio.run(test_lock())
```

if we run on separate terminals, first client 1 and then client 2 both end successfully but the resulting value is 1.

client 1
```
$ python c1.py 
Initial: None
Final: 1
```

client 2
```
$ python c2.py 
Initial: None
Final: 1
```

**Comment by ttecles** [comment]
@alisaifee I don't understand why we cannot implement the `immediate_execute_command` if as soon as you create a `ClusterPipelineImpl` you check that all keys are in the same node. As a consequence, any command to read data can be executed if the key is watched. You know what's been watched, and the node.

I don't understand why you are forcing all commands in a transaction to be executed on the same slot and we cannot run the `iimmediate_execute_command` to fetch the data in a ClusterPipeline. However, we are allowed to do it for a single Node.

**Comment by alisaifee** [comment]
>  I don't understand why we cannot implement the `immediate_execute_command` if as soon as you create a `ClusterPipelineImpl` you check that all keys are in the same node. As a consequence, any command to read data can be executed if the key is watched. You know what's been watched, and the node.
> 
> The code forces all commands in a transaction to be executed on the same slot but doesn't allow running the `immediate_execute_command` to fetch the data in a `ClusterPipelineImp`. However, we are allowed to do it for a single Node which is already forced on the `ClusterPipelineImp` to work in a single slot (i.e. Node)

My gut feel is the same from reading through the code properly. Keep in mind I'm not the original author of the pipeline implementation (it was mostly copied as is from aredis, with some code cleanup and improvements in type annotations).

Let me try to put together a minimal implementation today.

**Comment by alisaifee** [comment]
Keep in mind that this is a breaking change in the sense that `transaction` for the cluster client will now behave consistently with the standalone client. Therefore the instance of `pipe` that is passed to the callable `func` of `transaction` will start in "watching" mode and require a call to `multi` before the pipelined commands of the transaction can be issued.

Specifically, this will fail:
```py
async def test_lock():
    client = get_redis_cluster_client()

    async def client_side_incr(pipe):
        pipe = await pipe.get('KEY')
        current_value = (await pipe.execute())[0]
        print(f'Initial: {current_value}')
        current_value = int(current_value or 0)
        await pipe.set('KEY', current_value + 1)

    await client.transaction(client_side_incr, 'KEY')

    final_value = await client.get('KEY')
    print(f'Final: {final_value}')
```

but this will work the same way both with a cluster and standalone client:

```py
async def test_lock():
    client = get_redis_cluster_client()

    async def client_side_incr(pipe):
        current_value = await pipe.get('KEY')
        print(f'Initial: {current_value}')
        current_value = int(current_value or 0)
        pipe.multi()
        await pipe.set('KEY', current_value + 1)

    await client.transaction(client_side_incr, 'KEY')

    final_value = await client.get('KEY')
    print(f'Final: {final_value}')
```

**Comment by ttecles** [comment]
I tested it locally and it does work :partying_face: Thanks for the clarification that was the last thing we tried to see if we had a way of getting the data. But since WATCH works with MULTI/EXEC I see why we need to specify the `pipe.multi()` because before executing it we are not queueing commands.

**Comment by alisaifee** [comment]
Thanks for testing it out! I'll aim to do a release by EOD.

**Comment by alisaifee** [comment]
The fix is now  available in [4.9.0](https://github.com/alisaifee/coredis/releases/tag/4.9.0). Once you've verified it works for you please mark the issue closed.

**Comment by ttecles** [comment]
working! :partying_face:
